pest_meta/src/validator.rs

```diff
@@ -327,30 +311,30 @@ fn validate_whitespace_comment<'i>(rules: &Vec<ParserRule<'i>>) -> Vec<Error<'i,
         .collect()
 }
 
-fn validate_left_recursion<'i>(rules: &Vec<ParserRule<'i>>) -> Vec<Error<'i, GrammarRule>> {
+fn validate_left_recursion<'a, 'i: 'a>(rules: &'a Vec<ParserRule<'i>>) -> Vec<Error<'i, Rule>> {
     left_recursion(to_hash_map(rules))
 }
 
-fn to_hash_map<'a, 'i>(rules: &'a Vec<ParserRule<'i>>) -> HashMap<Ident, &'a ParserNode<'i>> {
+fn to_hash_map<'a, 'i: 'a>(rules: &'a Vec<ParserRule<'i>>) -> HashMap<&'i str, &'a ParserNode<'i>> {
     let mut hash_map = HashMap::new();
 
     for rule in rules {
-        hash_map.insert(rule.name.clone(), &rule.node);
+        hash_map.insert(rule.name, &rule.node);
     }
 
     hash_map
 }
 
-fn left_recursion<'i>(rules: HashMap<Ident, &ParserNode<'i>>) -> Vec<Error<'i, GrammarRule>> {
-    fn check_expr<'i>(
-        node: &ParserNode<'i>,
-        rules: &HashMap<Ident, &ParserNode<'i>>,
-        trace: &mut Vec<Ident>
-    ) -> Option<Error<'i, GrammarRule>> {
+fn left_recursion<'a, 'i: 'a>(rules: HashMap<&'i str, &'a ParserNode<'i>>) -> Vec<Error<'i, Rule>> {
```

Comment:
Does `'a` need to be introduced here? I would have imagined that using `'i` everywhere would work.

Reply:
The second lifetime is needed because the nodes can't live as long as the slices they reference. If we forced `&'i ParserNode<'i>` the nodes referenced in the HashMap would need to live at least as long as lifetime `'i`, which is also the lifetime of the slice that they reference. This, however, is not possible, since the node cannot live beyond the input as that would invalidate the references that it holds to the input.